Libplanet.Net/BlockCandidateTable.cs

```diff
@@ -18,79 +17,238 @@ namespace Libplanet.Net
+            _canonicalChainComparer = canonicalChainComparer;
+            _lock = new object();
         }
 
-        public bool Any() => !_blocks.IsEmpty;
+        /// <summary>
+        /// Retrieves the <see cref="CandidateBranch{T}"/> with the best Tip, i.e.
+        /// the highest <see cref="Block{T}.TotalDifficulty"/>.
+        /// If <see cref="BlockCandidateTable{T}.Update"/> is not called, this value is the best
+        /// branch in <see cref="BlockCandidateTable{T}.Branches"/> without any reduction.
+        /// </summary>
+        public CandidateBranch<T>? BestBranch { get; private set; }
+
+        public long Count => Branches.Count;
+
+        /// <summary>
+        /// An <see cref="IEnumerable{T}"/> of
+        /// <see cref="CandidateBranch{T}"/>es that can be appended
+        /// to the <em>current</em> <see cref="BlockChain{T}"/>.
+        /// </summary>
+        private ConcurrentBag<CandidateBranch<T>> Branches { get; set; }
 
         /// <summary>
-        /// Adds a <see cref="Block{T}"/>s to the table.
+        /// Adds an <see cref="CandidateBranch{T}"/> to <see cref="Branches"/>.
         /// </summary>
-        /// <param name="blockHeader">This is the header of the <see cref="BlockChain{T}"/>
-        /// tip at the time of downloading the blocks.</param>
-        /// <param name="blocks">List of downloaded <see cref="Block{T}"/>.</param>
-        public void Add(BlockHeader blockHeader, IEnumerable<Block<T>> blocks)
+        /// <param name="branch">The <see cref="CandidateBranch{T}"/>
+        /// to add to <see cref="Branches"/>.
+        /// </param>
+        /// <param name="currentTip">The topmost <see cref="BlockChain{T}.Tip"/> for
+        /// determining whether branch has higher comparer value.
+        /// </param>
+        public void Add(CandidateBranch<T> branch, IBlockHeader currentTip)
         {
-            if (_blocks.ContainsKey(blockHeader))
+            lock (_lock)
             {
-                return;
-            }
+                if (!IsBlockNeeded(branch.Tip, currentTip))
+                {
+                    return;
+                }
 
-            try
-            {
-                var sortedBlocks =
-                    new SortedList<long, Block<T>>(blocks.ToDictionary(i => i.Index));
-                _blocks.TryAdd(blockHeader, sortedBlocks);
-            }
-            catch (ArgumentException e)
-            {
-                Log.Debug(
-                    "Among the previous blocks of BlockHeader " +
-                    "#{Index} {BlockHash}, there is a block with the wrong index. " +
-                    "InnerMessage: {InnerMessage}",
-                    blockHeader.Index,
-                    blockHeader.Hash,
-                    e.Message);
+                // If BestBranch is not initialized, set incoming branch as BestBranch.
+                if (BestBranch is null)
+                {
+                    if (AreBlocksContinuous(branch.Blocks))
```

Comment:
Usually a check for sanity for an argument at the start of the method, if possible, makes more sense. Also, it would make even more sense if branch continuity is checked by `CandidateBranch<T>` during its creation so we can remove `AreBlocksContinuous` method. 😛

Reply:
https://github.com/planetarium/libplanet/blob/2d826486ed23614e840c6b14e0d89c0e8c7da084/Libplanet.Net/CandidateBranch.cs#L22-L55

Now the block order sanity will be checked when a `CandidateBranch<T>` is made. 
And `CandidateBranch<T>` will be empty if the blocks are not in order.